app/services/PaypalService.scala

```diff
@@ -88,34 +97,28 @@ class PaypalService(config: PaypalApiConfig, contributionData: ContributionData)
 
     val payment = new Payment().setIntent("sale").setPayer(payer).setTransactions(transactions).setRedirectUrls(redirectUrls)
 
-    Try {
-      val createdPayment = payment.create(apiContext)
-      createdPayment.getLinks.asScala
-    } match {
-      case Success(links) =>
-        val approvalLink = links.find(_.getRel.equalsIgnoreCase("approval_url")).map(l => Right(addUserActionParam(l.getHref)))
-        approvalLink.getOrElse(Left("No approval link returned from paypal"))
-      case Failure(exception) => Left(exception.getMessage)
+    val res = asyncExecute {
+      val links = payment.create(apiContext).getLinks.asScala
+      links.find(_.getRel.equalsIgnoreCase("approval_url")).map(l => addUserActionParam(l.getHref))
```

Comment:
I think you need to separate your concerns here. You're both searching for something and mapping on the value.

I would wrap only `payment.create(apiContext)` in the async block and start from there